fury/shaders/base.py

```diff
@@ -0,0 +1,136 @@
+    ----------
+    actor : vtkActor
+        Object where you want to add the shader code.
+    shader_type : str
+        Shader type: vertex, geometry, fragment
+    impl_code : str, optional
+        shader implementation code, should be a string or filename
+    decl_code : str, optional
+        shader declaration code, should be a string or filename
+        by default None
+    block : str, optional
+        section name to be replaced, by default
+    keep_default : bool, optional
+        keep the default block tag to let VTK replace it with its default
+        behavior. By default True
+    replace_first : bool, optional
+        If True, apply this change before the standard VTK replacements
+        by default True
+    replace_all : bool, optional
+        [description], by default False
+    debug : bool, optional
+        introduce a small error to debug shader code.
+        by default False
+
+    """
+    shader_type = shader_type.lower()
+    shader_type = SHADERS_TYPE.get(shader_type, None)
+    if shader_type is None:
+        msg = "Invalid Shader Type. Please choose between "
+        msg += ', '.join(SHADERS_TYPE.keys())
+        raise ValueError(msg)
+
+    block = block.lower()
+    block = SHADERS_BLOCK.get(block, None)
+    if block is None:
+        msg = "Invalid Shader Type. Please choose between "
+        msg += ', '.join(SHADERS_TYPE.keys())
+        raise ValueError(msg)
+
+    block_dec = block + "::Dec"
+    block_impl = block + "::Impl"
+
+    if keep_default:
+        decl_code = block_dec + "\n" + decl_code
+        impl_code = block_impl + "\n" + impl_code
+
+    if debug:
+        error_msg = "\n\n--- DEBUG: THIS LINE GENERATES AN ERROR ---\n\n"
+        impl_code += error_msg
+
+    sp = actor.GetShaderProperty() if VTK_9_PLUS else actor.GetMapper()
+
+    sp.AddShaderReplacement(shader_type, block_dec, replace_first,
+                            decl_code, replace_all)
+    sp.AddShaderReplacement(shader_type, block_impl, replace_first,
+                            impl_code, replace_all)
+
+
+def add_shader_callback(actor, callback):
+    """Add a shader callback to the actor.
+
+    Parameters
+    ----------
+    actor : vtkActor
+        Rendered Object
+    callback : callable
+        function or class that contains 3 parameters: caller, event, calldata.
+        This callback will be trigger at each `UpdateShaderEvent` event.
+
+    """
+    @vtk.calldata_type(vtk.VTK_OBJECT)
+    def cbk(caller, event, calldata=None):
+        callback(caller, event, calldata)
+
+    mapper = actor.GetMapper()
+    mapper.AddObserver(vtk.vtkCommand.UpdateShaderEvent, cbk)
+
+
+def add_array_as_vertex_attribute(actor, arr, arr_name, attr_name, deep=True):
+    """Link a numpy array with vertex attribute.
+
+    Parameters
+    ----------
+    actor : vtkActor
+        Rendered Object
+    arr : ndarray
+        array to link to vertices
+    arr_name : str
+        data array name
+    attr_name : str
```

Comment:
Could we just keep one of the 2 params names in the signature of the function and if we need to differentiate between the 2 names do it by appending a fixed prefix to one of the 2 strings. My suggestion would be to leave the users to define the attribute they have to retrieve in the vertex shader and leave this function to manage the name of the vtk array.

Reply:
👍